functions/src/modules/gmail.functions.ts

```diff
@@ -0,0 +1,273 @@
+import * as functions from 'firebase-functions';
+import * as admin from 'firebase-admin';
+import * as Google from 'googleapis';
+import * as querystring from 'querystring';
+import { Message } from 'firebase-functions/lib/providers/pubsub';
+
+const db = admin.database();
+
+export const oauth2init = functions.https.onRequest(
+  async (req: functions.Request, res: functions.Response) => {
+    const { client_key, secret, redirect } = functions.config().gmail;
+    const oauth2Client = new Google.google.auth.OAuth2(
+      client_key,
+      secret,
+      redirect
+    );
+    const authURL = oauth2Client.generateAuthUrl({
+      access_type: 'offline',
+      scope: ['https://www.googleapis.com/auth/gmail.modify'],
+      prompt: 'consent',
+    });
+    res.redirect(authURL);
+  }
+);
+
+const saveTokens = async (values: any) => {
+  const newGmailTokensRef = db.ref(`/gmail/${values.emailKey}`);
```

Comment:
I have changed the card description in regard to the database path for this. Please update.

Reply:
Removed the email from path.